packages/slice-machine/src/features/onboarding/helpers/useOnboardingProgress.ts

```diff
@@ -0,0 +1,70 @@
+import { Dispatch, SetStateAction, useEffect, useRef } from "react";
+
+import { onboardingSteps as steps } from "@/features/onboarding/content";
+import type {
+  OnboardingStep,
+  OnboardingStepStatuses,
+  OnboardingStepType,
+} from "@/features/onboarding/types";
+import { useLocalStorageItem } from "@/hooks/useLocalStorageItem";
+
+const getInitialState = (): OnboardingStepStatuses => {
+  // if the old guide was dismissed, all steps start as complete
+  const startComplete =
+    localStorage.getItem("slice-machine_isInAppGuideOpen") === "false";
+
+  return Object.fromEntries(
+    steps.map((step) => [step.id, startComplete]),
+  ) as OnboardingStepStatuses;
+};
+
+const useOnboardingStepStatus = (): [
+  OnboardingStepStatuses,
+  Dispatch<SetStateAction<OnboardingStepStatuses>>,
+] => {
+  const initialState = useRef(getInitialState()).current;
+  const [status, setStatus, { isUnset: isStatusUnset }] = useLocalStorageItem(
+    "onboardingSteps",
+    initialState,
+  );
+
+  useEffect(() => {
+    // populate onboarding status if not defined
+    if (isStatusUnset) setStatus(initialState);
+
+    /* eslint-disable-next-line react-hooks/exhaustive-deps -- 
+    We don't care if the values of isStatusUnset, setStatus or initialState change here. */
+  }, [status]);
+
+  return [status, setStatus];
+};
```

Comment:
I would usually never agree with `never disable the exhaustive deps rule`, but you made a valid point about adjusting the state during render time, so that sound like a good solution to me 👀